metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,208 @@
+      //TODO: Validate transactions PM-3131/3132
+      txsValid <- Some(false).pure[F]
+    } yield txsValid.map(_ && ledgerHashValid)
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Boolean] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+
+    if (commitQC.blockHash != block.hash) {
+      lastExecutedHeader.get
+        .flatMap { h =>
+          ledgerUpdateSemaphore.withPermit(
+            updateLedger(block, h.height)
+          )
+        }
+        .as(false)
+    } else {
+      Sync[F]
+        .bracket {
+          ledgerUpdateSemaphore.acquireN(maxConcurrentUpdates)
+        } { _ =>
+          for {
+            lastExec <- lastExecutedHeader.get
+            ledger   <- updateLedger(block, lastExec.height)
+            chkpOpt  <- constructCheckpoint(ledger, commitQC)
+            _        <- commitState(block.header, ledger)
+            -        <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
+          } yield false
+        } { _ =>
+          ledgerUpdateSemaphore.releaseN(maxConcurrentUpdates)
+        }
+    }
+  }
+
+  private def updateLedger(
+      block: Block,
+      executedHeight: Long
+  ): F[Ledger] = {
+
+    def loop(block: Block): F[Ledger] = {
+
+      getLedger(block.header.parentHash).flatMap {
+        case Some(ledger) =>
+          updateLedgerByBlock(ledger, block)
+
+        case None if block.header.height <= executedHeight =>
+          Sync[F].raiseError(
+            new IllegalStateException(
+              "Attempting to execute a branch that does not extend last executed block. Aborting execution"
+            )
+          )
+
+        case None =>
+          getBlock(block.header.parentHash).flatMap {
+            case Some(parent) =>
+              loop(parent) >>= (updateLedgerByBlock(_, block))
+
+            case None =>
+              Sync[F].raiseError(
+                new IllegalStateException(s"Block not available: ${block.hash}")
+              )
+          }
+      }
+
+    }
+
+    OptionT(getLedger(block.hash)).getOrElseF(loop(block))
+  }
+
+  private def updateLedgerByBlock(
+      oldLedger: Ledger,
+      block: Block
+  ): F[Ledger] = {
+    val newLedger = block.body.transactions.foldLeft(oldLedger)(_ update _)
+    ledgerTree
+      .update {
+        _ + (block.hash -> (newLedger, block.header))
+      }
+      .as(newLedger)
+  }
+
+  private def getBlock(hash: Block.Hash): F[Option[Block]] =
+    storeRunner.runReadOnly(blockStorage.get(hash))
+
+  private def getLedger(hash: Block.Hash): F[Option[Ledger]] =
+    ledgerTree.get.map(_.get(hash).map(_._1))
+
+  private def commitState(header: Block.Header, ledger: Ledger): F[Unit] = {
+    storeRunner.runReadWrite {
+      ledgerStorage.put(ledger) >>
+        viewStateStorage.setLastExecutedBlockHash(header.hash)
```

Comment:
It would be best not to modify the last executed block hash here:
1. You will return a flag to indicate whether this should be done by the executor
2. The state sync mechanism may have done it in the meantime, which is something the executor will check before modifying it

Reply:
I did not consider state sync. But I wonder isn't it the more reason to update the last executed hash here? We could also use `compareAndSetLastExecutedBlockHash` here and if we detect that it has changed outside of the execution we would update the `ledgerTree` accordingly. This would be just an optimisation though, the CheckpointingService should do fine if the internal in-memory state lags behind.

Reply:
Ah, I guess `CheckpointingService#syncState` will handle that anyway, right?

Reply:
Yes, the `CheckpointingService.syncState` would be called when it's being asked to jump ahead and it can re-establish a correct `ledgerTree`. 

The reason I think to keep it out of here is that if you had to explain to someone how to implement an `ApplicationService` it would be best not to have to explain things like "we also expect you to update this storage when X and Y happens". i thought it was best to make these requests part of the return values from functions, so we can take care of them in the common parts.